tests/test_settings.py

```diff
@@ -128,6 +130,62 @@ def test_base_path_overrides_config_file_code_and_deps(
     assert settings.deps == expected
 
 
+OPTION_VALUES = {
+    "code": ["a", "b", "c"],
+    "deps": ["d", "e"],
+    "ignore-undeclared": ["f", "g"],
+    "ignore-unused": ["h", "i", "j"],
+}
+
+
+def multivalued_optargs_grid() -> List[List[str]]:
+    """Create command-line option/argument combinations which
+    mix order and number of multivalued parameters."""
+
+    def powerset_non_empties(iterable):
+        xs = list(iterable)
+        return chain.from_iterable(combinations(xs, k) for k in range(1, len(xs)))
+
+    opt_by_arg = {arg: opt for opt, argvals in OPTION_VALUES.items() for arg in argvals}
+    parameter_grids = []
+    for items in OPTION_VALUES.values():
+        split_2 = set(
+            chain.from_iterable(
+                [(x, y), (y, x)]
+                for x, y in combinations(powerset_non_empties(items), 2)
+                if len(x + y) == len(items) and not set(x) & set(y)
+            )
+        )
+        parameter_grids.append(split_2)
```

Comment:
If we look at the `split_2` set as all the pairs that show a different way to order + cover the chars in the list, then we can rephrase its construction as generating permutations of all the values, and then finding ways to slice those permutations into non-empty pairs. Consider this helper:

```
    parameter_grids = [
        set(chain.from_iterable(map(slice_in_two, permutations(items))))
        for items in OPTION_VALUES.values()
    ]
```
with this helper:
```python
    def slice_in_two(items: List[T]) -> Iterator[Tuple[List[T], List[T]]]:
        assert len(items) >= 2
        for i in range(1, len(items)):
            yield items[:i], items[i:]
```

or using comprehensions instead of `map` and `chain.from_iterable`:

```suggestion
    def slice_in_two(items: List[T]) -> Iterator[Tuple[List[T], List[T]]]:
        assert len(items) >= 2
        for i in range(1, len(items)):
            yield items[:i], items[i:]

    parameter_grids = [
        {
            pair
            for sequence in permutations(items)
            for pair in slice_in_two(sequence)
        }
        for items in OPTION_VALUES.values()
    ]
```

Reply:
OK, i've gone with an implementation based on this idea here: 400e34a7330944499435ce1a63439179a68d6ca2